File: atcoder/TypicalContest/Contest001/A.py
from typing import List
import sys

sys.setrecursionlimit(10 ** 6)

H, W = map(int, input().split())


class Block:
    def __init__(self, val: str, x: int, y: int):
        self.val = val
        self.x = x
        self.y = y

    def can_enter(self) -> bool:
        return self.val in ["s", "g", "."]


fields: List[List[Block]] = [[None for i in range(W)] for j in range(H)]

dxs = [-1, 0, 0, 1]
dys = [0, 1, -1, 0]

start = None
goal = None

for i in range(H):
    blocks = list(map(str, input()))
    for j, c in enumerate(blocks):
        block = Block(c, j, i)
        if c == "s":
            start = block
        elif c == "g":
            goal = block
        fields[i][j] = block

seen = [[False for i in range(W)] for j in range(H)]

ans = False


def dfs(edge: Block):
    global ans
    seen[edge.y][edge.x] = True
    if ans:
        return

    if edge.val == "g":
        ans = True
        return

    for dx, dy in zip(dxs, dys):
        x = edge.x + dx
        y = edge.y + dy
        if x < 0 or x >= W:
            continue
        if y < 0 or y >= H:
            continue
        if seen[y][x]:
            continue
        nxt = fields[y][x]
        if not nxt.can_enter():
            continue
        dfs(nxt)


dfs(start)
print("Yes" if ans else "No")

File: atcoder/TypicalContest/Contest001/A.cpp
#include <iostream>
#include <vector>

using namespace std;

struct Block
{
    int x;
    int y;
    char c;

    bool can_enter()
    {
        return c != '#';
    }
};

using Graph = vector<vector<Block>>;

vector<vector<bool>> seen;
vector<vector<Block>> fields;

vector<int> dxs{1, -1, 0, 0};
vector<int> dys{0, 0, -1, 1};

Block sb;
Block gb;

void dfs(const Block &edge, int H, int W)
{
    seen[edge.y][edge.x] = true;
    for (int i = 0; i < 4; i++)
    {
        int nx = edge.x + dxs[i];
        int ny = edge.y + dys[i];
        if (nx < 0 || nx >= W || ny < 0 || ny >= H)
        {
            continue;
        }
        if (!fields[edge.y][edge.x].can_enter())
        {
            continue;
        }
        if (seen[ny][nx])
        {
            continue;
        }
        Block nxt = fields[ny][nx];
        dfs(nxt, H, W);
    }
}

int main()
{
    int H, W;
    cin >> H >> W;
    fields.assign(H, vector<Block>(W));
    seen.assign(H, vector<bool>(W, false));
    for (int i = 0; i < H; i++)
    {
        string line;
        cin >> line;
        for (int j = 0; j < line.length(); j++)
        {
            Block block{j, i, line[j]};
            if (block.c == 's')
            {
                sb = block;
            }
            else if (block.c == 'g')
            {
                gb = block;
            }
            fields[i][j] = block;
        }
    }
    dfs(sb, H, W);
    if (seen[gb.y][gb.x])
    {
        cout << "Yes";
    }
    else
    {
        cout << "No";
    }

    return 0;
}